File: typescript/src/key-pair.ts
import {
  autograph_ephemeral_key_pair,
  autograph_identity_key_pair,
  autograph_key_pair_size
} from './clib'

const createKeyPair = () => new Uint8Array(autograph_key_pair_size())

export const generateIdentityKeyPair = (): Uint8Array => {
  const keyPair = createKeyPair()
  const success = autograph_identity_key_pair(keyPair)
  if (!success) {
    throw new Error('Key generation failed')
  }
  return keyPair
}

export const generateKeyPair = (): Uint8Array => {
  const keyPair = createKeyPair()
  const success = autograph_ephemeral_key_pair(keyPair)
  if (!success) {
    throw new Error('Key generation failed')
  }
  return keyPair
}

File: rust/src/key_pair.rs
use rand_core::{CryptoRng, RngCore};

use crate::{
    constants::KEY_PAIR_SIZE,
    error::Error,
    external::{init, key_pair_ephemeral, key_pair_identity},
    types::KeyPair,
};

fn ephemeral_key_pair<T: RngCore + CryptoRng>(csprng: T, key_pair: &mut KeyPair) -> bool {
    if !init() {
        return false;
    }
    key_pair_ephemeral(csprng, key_pair)
}

fn identity_key_pair<T: RngCore + CryptoRng>(csprng: T, key_pair: &mut KeyPair) -> bool {
    if !init() {
        return false;
    }
    key_pair_identity(csprng, key_pair)
}

pub fn generate_key_pair<T: RngCore + CryptoRng>(csprng: T) -> Result<KeyPair, Error> {
    let mut key_pair: KeyPair = [0; KEY_PAIR_SIZE];
    let success = ephemeral_key_pair(csprng, &mut key_pair);
    if !success {
        Err(Error::KeyPair)
    } else {
        Ok(key_pair)
    }
}

pub fn generate_identity_key_pair<T: RngCore + CryptoRng>(csprng: T) -> Result<KeyPair, Error> {
    let mut key_pair: KeyPair = [0; KEY_PAIR_SIZE];
    let success = identity_key_pair(csprng, &mut key_pair);
    if !success {
        Err(Error::KeyPair)
    } else {
        Ok(key_pair)
    }
}
